servers/src/main/java/tachyon/worker/file/FileDataManager.java

```diff
@@ -72,13 +79,90 @@ public FileDataManager(BlockDataManager blockDataManager) {
   }
 
   /**
+   * Checks if the given file is being persisted.
+   *
+   * @param fileId the file id
+   * @return true if the file is being persisted, false otherwise
+   */
+  private boolean isFilePersisting(long fileId) {
+    synchronized (this) {
+      return mPersistedFiles.contains(fileId);
+    }
+  }
+
+  /**
+   * Checks if the given file needs persistence.
+   *
+   * @param fileId the file id
+   * @return false if the file is being persisted, or is already persisted; otherwise true
+   */
+  public boolean needPersistence(long fileId) {
+    if (isFilePersisting(fileId) || isFilePersisted(fileId)) {
+      return false;
+    }
+
+    try {
+      if (fileExistsInUfs(fileId)) {
+        // mark as persisted
+        addPersistedFile(fileId);
+        return false;
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to check if file {} exists in under storage system", fileId, e);
+    }
+    return true;
+  }
+
+  /**
+   * Checks if the given file is persisted.
+   *
+   * @param fileId the file id
+   * @return true if the file is being persisted, false otherwise
+   */
+  public boolean isFilePersisted(long fileId) {
+    synchronized (this) {
+      return mPersistedFiles.contains(fileId);
+    }
+  }
+
+  /**
+   * Adds a file as persisted.
+   *
+   * @param fileId the file id
+   */
+  private void addPersistedFile(long fileId) {
+    synchronized (this) {
+      mPersistedFiles.add(fileId);
+    }
+  }
+
+  /**
+   * Checks if the given file exists in the under storage system.
+   *
+   * @param fileId the file id
+   * @return true if the file exists in under storage system, false otherwise
+   * @throws IOException an I/O exception occurs
+   */
+  public synchronized boolean fileExistsInUfs(long fileId) throws IOException {
+    String ufsRoot = mTachyonConf.get(Constants.UNDERFS_ADDRESS);
+    FileInfo fileInfo = mBlockDataManager.getFileInfo(fileId);
+    String dstPath = PathUtils.concatPath(ufsRoot, fileInfo.getPath());
+
+    return mUfs.exists(dstPath);
+  }
```

Comment:
It's called by `needPersistence`, which is called before the file is scheduled. Also, this class can be private.